src/main/java/com/es/phoneshop/service/impl/RecentlyViewedProductServiceImpl.java

```diff
@@ -0,0 +1,63 @@
+package com.es.phoneshop.service.impl;
+
+import com.es.phoneshop.model.Product;
+import com.es.phoneshop.model.RecentlyViewedProduct;
+import com.es.phoneshop.service.ProductService;
+import com.es.phoneshop.service.RecentlyViewedProductService;
+import jakarta.servlet.http.HttpSession;
+
+import java.util.Set;
+import java.util.UUID;
+
+public class RecentlyViewedProductServiceImpl implements RecentlyViewedProductService {
+    private static final String RECENTLY_VIEWED_PRODUCT_SESSION_ATTRIBUTE = RecentlyViewedProductServiceImpl.class.getName() +
+            ".product";
+    private static final int MAX_RECENTLY_VIEWED_PRODUCTS = 3;
+    private final ProductService productService;
+
+    private RecentlyViewedProductServiceImpl() {
+        this.productService = new ProductServiceImpl();
+    }
+
+    private static final class SingletonHolder {
+        private static final RecentlyViewedProductServiceImpl INSTANCE = new RecentlyViewedProductServiceImpl();
+    }
+
+    public static RecentlyViewedProductServiceImpl getInstance() {
+        return SingletonHolder.INSTANCE;
+    }
+
+    @Override
+    public Set<Product> get(HttpSession session) {
+        synchronized (session.getId()) {
+            RecentlyViewedProduct viewedProducts = (RecentlyViewedProduct) session
+                    .getAttribute(RECENTLY_VIEWED_PRODUCT_SESSION_ATTRIBUTE);
+
+            if (viewedProducts == null) {
+                viewedProducts = new RecentlyViewedProduct();
+                session.setAttribute(RECENTLY_VIEWED_PRODUCT_SESSION_ATTRIBUTE, viewedProducts);
+            }
+            return viewedProducts.getProducts();
+        }
+    }
+
+    @Override
+    public void add(UUID productId, HttpSession session) {
+        synchronized (session.getId()) {
+            Product product = productService.getProduct(productId);
+            RecentlyViewedProduct viewedProducts = (RecentlyViewedProduct) session
+                    .getAttribute(RECENTLY_VIEWED_PRODUCT_SESSION_ATTRIBUTE);
+            Set<Product> products = viewedProducts.getProducts();
```

Comment:
if you use set it means that the order of items is not considered at all
Is it expected behavior?